Purchase Order tracking

chiaberry opened the pull request:
## Associated issues

https://github.com/cityofaustin/atd-data-tech/issues/8259

## Description

Adds a new moped_purchase_orders table to the database, as well as putting the UI on the funding tab. 

![Screen Shot 2022-07-21 at 11 56 19 AM](https://user-images.githubusercontent.com/12474808/180274547-27b6c2c7-1c50-42e9-9e25-0a3ad25acebb.png)

There is also a script in the moped-toolbox to get the contractors and DOs from existing records. Heavily cribbed from John's milestone backfill script. 

I have left the contractor and DO fields on the project summary page for now, After this is approved, I will open a new PR to remove those fields  (https://github.com/cityofaustin/atd-data-tech/issues/9770). The issue I wrote says to remove the fields and drop the columns, but now I'm thinking that we will have to wait to drop the columns in a subsequent release (not the S release but the T release) since we wont be able to run the backfill script in prod until after the S release. 

## Testing
**URL to test:** <!-- https://moped-test.austinmobility.io/moped/ or Netlify -->

Run locally or moped test
https://moped-test.austinmobility.io/moped/projects/161?tab=funding

**Steps to test:**

to test the purchase order table: 
- add / edit / delete / update records to the table on the funding tab. 

to test the script:
- to test locally, I added some contractors and DOs via the summary tab to various projects.
- Then I ran the script (`create_purchase_orders.py -e local`)
- And checked on the funding tab that the information I added in the summary tab showed up on the PO table

---
#### Ship list
- [x] Code reviewed 
- [ ] Product manager approved
- [ ] Added to [QA test script](https://docs.google.com/spreadsheets/d/1n_O6MLh9cwwPf57HUM394Ea-z9uuoEV1-QW4axNZXLE/edit#) if applicable

frankhereford commented on moped-toolbox/purchase_order_backfill/create_purchase_orders.py:
```diff
@@ -0,0 +1,71 @@
+import argparse
+
+
+from amd_milestones_backfill.utils import get_logger, make_hasura_request
+from queries import PROJECTS_QUERY, ADD_PURCHASE_ORDER
+from amd_milestones_backfill.secrets import HASURA
```
I recognize that these are imports from a sibling toolbox folder, but because of that, I'm not able to get the script to run unmodified. I can symlink some stuff around to make it work, but I can't imagine that's what is intended. Please let me know if I'm missing something obvious 🙏

chiaberry replied:
Not missing anything obvious at all, I struggled with the imports. I decided to include the files in the same folder also keeping the different tools in the toolbox more isolated from each other.

frankhereford reviewed (commented):
Do you think the page should reject the creation of a PO record when none of the three fields are given data? I'm on the fence; it making an empty record when I submit an empty record feels right, but would that lead us to accumulating unused, distracting, empty records?

Do you think that a very brief entry in the toolbox `README.md` file to describe what the tool does (and considerations for use?) would be helpful? You may also want to add some instructions for getting it running.

chiaberry commented:
@frankhereford thank you for the thoughtful feedback, I've added instructions and a summary to the readme and updated the files so the script runs easily from the command line. 

I think it's a good question about if we should prevent users from saving an empty purchase order record, what does everyone else think?

frankhereford reviewed (approved):
Thanks @chiaberry!

patrickm02L reviewed (approved):
I couldn't break it. 🪨 solid

https://user-images.githubusercontent.com/5566727/180868224-4ecedae7-6466-4b73-b6fe-319e83dd620d.mp4

!

mddilley commented on moped-editor/src/views/projects/projectView/ProjectPurchaseOrderTable.js:
```diff
@@ -0,0 +1,256 @@
+
+const ProjectPurchaseOrderTable = () => {
+  // addAction Ref - mutable ref object used to access add action button imperatively.
+  const addActionRef = React.useRef();
```
I think this ref can go since it isn't used to access the add button anywhere. I could be missing something though

mddilley reviewed (approved):
I tested the table and the backfill script and all works just as promised. 🎉 Thanks for the docs  in the 🐍 script - it made testing the backfill a breeze! 🙏

frankhereford reviewed (changes requested):
I think this looks great. I've tested it locally and gotten the backfill script to run. I put it in a `venv` and then installed the requests package to do that. Do you think a `requirements.txt` would be a helpful addition in that toolbox directory? I think it would only need one line in it:

```
requests==2.*
```

Thanks!!

frankhereford commented on moped-toolbox/README.md:
```diff
@@ -37,3 +38,17 @@ to backfill project milestones after implementing the milestone template feature
+
+### Usage instructions
+
+1. Create a python 3.x environment with `requests` installed.
```
I wanted to recognize that you do have all of the instructions one could need to get this running even though I mentioned a requirements.txt in my comment above. 🙏

johnclary commented on moped-editor/src/views/projects/projectView/ProjectPurchaseOrderTable.js:
```diff
@@ -0,0 +1,256 @@
+    open: false,
+    message: null,
+    severity: "success",
+  };
```
nitpicking, but you can bump `DEFAULT_SNACKBAR_STATE` outside of the component

johnclary commented on moped-toolbox/purchase_order_backfill/create_purchase_orders.py:
```diff
@@ -0,0 +1,70 @@
+"""
```
looks good! just noting that if something goes awry in the midst of running this, it won't be easy to pickup where you left off. will need to just purge all the purchase_order records and start over. not a big deal, but we should run this as soon as it's deployed to prod. once users start adding their own purchase order records it will be hard to untangle.

johnclary reviewed (approved):
looks great; works great. awesome! 🚢

mateoclarke reviewed (approved):
🚢 💨 ✨

johnclary commented:
i removed the `ref` and moved the default snackbar state.

mateoclarke commented:
Do we need to add a line to the Deploy Notes section of the Release Notes so we don't forget to run the PO backfill script?

johnclary commented:
@amenity @chiaberry FYI i added this to the test script and removed the old test steps from the "Summary" section.
